feat: add `assert/is-well-formed-string`

gunjjoshi opened the pull request:
This commit adds the assert/is-well-formed-string module for checking well-formed-strings.

Resolves #1065 

## Description

Added the is-well-formed-string package. It checks for four things to check if a string is well formed or not:

  1. A low surrogate is present at the beginning.
  2. A high surrogate is present at the last.
  3. No low surrogate after a high surrogate.
  4. No high surrogate before a low surrogate.

> What is the purpose of this pull request?

This pull request:

-   Adds a package to check if a string is well-formed or not.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1065 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Screenshot of tests:

<img width="367" alt="Screenshot 2024-02-26 at 10 53 57" src="https://github.com/stdlib-js/stdlib/assets/67432819/bc4d0c60-7c55-4dd2-ac80-c4e610f80c2e">

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   ☑️  Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gunjjoshi committed:
feat: added string/is-well-formed-string

This commit adds the string/is-well-formed-string module for checking well-formed-strings.

gunjjoshi committed:
feat: added assert/is-well-formed-string

This commit adds the assert/is-well-formed-string module for checking well-formed-strings.

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/assert/is-negative-number",
+  "version": "0.0.0",
+  "description": "Test if a value is a number having a negative value.",
```
@gunjjoshi Could you please update the name, description, and keywords in the `package.json`?

gunjjoshi replied:
Sure, I've made the suggested changes.

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/primitive.js:
```diff
@@ -0,0 +1,51 @@
+*/
+function isWellFormedString( str ) {
+	return (
+		typeof str === 'string' && isWellFormed(str)
```
```suggestion
		typeof str === 'string' && isWellFormed( str )
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/README.md:
```diff
@@ -0,0 +1,139 @@
+bool = isWellFormedString( '\uDBFF' );
+// returns false
+bool = isWellFormedString( null );
+// returns false
```
```suggestion
var bool = isWellFormedString( '' );
// returns true

bool = isWellFormedString( new String( '' ) );
// returns true

bool = isWellFormedString( '\uD800\uD800' );
// returns false

bool = isWellFormedString( '\uDBFF' );
// returns false

bool = isWellFormedString( null );
// returns false
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/README.md:
```diff
@@ -0,0 +1,139 @@
+var bool = isWellFormedString.isObject( '' );
+// returns false
+bool = isWellFormedString.isObject( new String( '' ) );
+// returns true
```
```suggestion
var bool = isWellFormedString.isObject( '' );
// returns false

bool = isWellFormedString.isObject( new String( '' ) );
// returns true
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/README.md:
```diff
@@ -0,0 +1,139 @@
+bool = isWellFormedString( '-5' );
+// returns true
+bool = isWellFormedString( null );
+// returns false
```
```suggestion
var isWellFormedString = require( '@stdlib/assert/is-well-formed-string' );

var bool = isWellFormedString( '' );
// returns true

bool = isWellFormedString( new String( '' ) );
// returns true

bool = isWellFormedString( '\uDBFF' );
// returns false

bool = isWellFormedString( '\uDBFF\uDBFF' );
// returns false

bool = isWellFormedString( [] );
// returns false

bool = isWellFormedString( '-5' );
// returns true

bool = isWellFormedString( null );
// returns false
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/README.md:
```diff
@@ -0,0 +1,139 @@
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/assert/is-String`][@stdlib/assert/is-String]</span><span class="delimiter">: </span><span class="description">test if a str is a String.</span>
```
No reason to manually populate this section; we have tooling that automatically takes care of it.

```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/README.md:
```diff
@@ -0,0 +1,139 @@
+
+<!-- <related-links> -->
+
+[@stdlib/assert/is-string]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/assert/is-string
```
```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts:
```diff
@@ -0,0 +1,61 @@
+}
+
+
+// Attached to main export is an isPrimitive method which returns a boolean...
```
```suggestion
// Attached to main export is an isObject method which returns a boolean...
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts:
```diff
@@ -0,0 +1,61 @@
+	isWellFormedString.isPrimitive(); // $ExpectError
+	isWellFormedString.isPrimitive( '', '\uDBFF' ); // $ExpectError
+}
+
```
```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js:
```diff
@@ -0,0 +1,44 @@
+// => true
+
+console.log( isWellFormedString( new String( '' ) ) );
+// => false
```
Shouldn't this be `true`? Am confused why linting didn't pick on this...

```suggestion
// => true
```

gunjjoshi replied:
Yes, it should be true.

Planeshifter replied:
Thanks for confirming!

Planeshifter reviewed (changes requested):
Thanks for working on this! Looks very good; left some comments that will need to be addressed before merging.

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/README.md

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+* // returns false
+*/
+function isWellFormed(str) {
+	for ( i = 0; i < str.length; i++) {
```
```suggestion
	for ( i = 0; i < str.length; i++ ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+			return false;
+		}
+
+		// Checking if a high surrogate is present at the last
```
```suggestion
		// Checking if a high surrogate is present at the last position
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+		}
+
+		// Checking if there is no low surrogate after a high surrogate
+		if (i < str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
```
```suggestion
		if ( i < str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i]) ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+
+		// Checking if there is no low surrogate after a high surrogate
+		if (i < str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
+			if (RE_UTF16_LOW_SURROGATE.test(str[i + 1])) {
```
```suggestion
			if ( RE_UTF16_LOW_SURROGATE.test(str[i + 1]) ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+		// Checking if there is no low surrogate after a high surrogate
+		if (i < str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
+			if (RE_UTF16_LOW_SURROGATE.test(str[i + 1])) {
+				i+=1;
```
```suggestion
				i += 1;
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+		}
+
+		// Checking if there is no high surrogate before a low surrogate
+		if ( i>0 && RE_UTF16_LOW_SURROGATE.test(str[i])) {
```
```suggestion
		if ( i > 0 && RE_UTF16_LOW_SURROGATE.test(str[i]) ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+
+		// Checking if there is no high surrogate before a low surrogate
+		if ( i>0 && RE_UTF16_LOW_SURROGATE.test(str[i])) {
+			if (!RE_UTF16_HIGH_SURROGATE.test(str[i - 1])) {
```
```suggestion
			if ( !RE_UTF16_HIGH_SURROGATE.test(str[i - 1]) ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+
+var RE_UTF16_LOW_SURROGATE = /[\uDC00-\uDFFF]/;
+var RE_UTF16_HIGH_SURROGATE = /[\uD800-\uDBFF]/;
+var i = 0;
```
We should not initialize the `i` variable outside the function scope. Just declare `var i;` in L45 before the loop.

gunjjoshi replied:
I assumed that all variables should be declared in the variables section. 
But yes, it is outside the scope of the function, and we should avoid using global variables.
I've made the suggested change.

gunjjoshi committed:
Update package.json

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/primitive.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/README.md

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/README.md

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/README.md

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/README.md

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update iswellformed.js

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update iswellformed.js

Signed-off-by: GUNJ JOSHI <gunjjoshi8372@gmail.com>

gunjjoshi commented:
@Planeshifter 
There is some error in linting the examples. I have written the correct outputs, but it is expecting something else, and thus giving errors. I checked the implementation of the functions, it is indeed correct. 

<img width="1710" alt="Screenshot 2024-02-27 at 10 36 46" src="https://github.com/stdlib-js/stdlib/assets/67432819/6eaffedc-eb58-43ec-bde8-717688e19ada">

Planeshifter commented:
@gunjjoshi Thanks for checking! I believe the issue is that the `instanceof String` check doesn't work across realms when there are multiple `String` constructors. To lint examples and their expected outputs, we spawn a separate process via `vm2` that will have it's own `String` constructor. Here is how we make the object check in the `assert/is-string` package https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/assert/is-string/lib/object.js more robust to have it work across realms.
Would you mind updating this package accordingly?

gunjjoshi commented:
> @gunjjoshi Thanks for checking! I believe the issue is that the `instanceof String` check doesn't work across realms when there are multiple `String` constructors. To lint examples and their expected outputs, we spawn a separate process via `vm2` that will have it's own `String` constructor. Here is how we make the object check in the `assert/is-string` package https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/assert/is-string/lib/object.js more robust to have it work across realms. Would you mind updating this package accordingly?

Sure, I'll update this package.

gunjjoshi committed:
feat: added assert/is-well-formed-string

This commit adds the assert/is-well-formed-string module for checking well-formed-strings.

gunjjoshi commented:
Great ! It got passed by the CI pipeline.
So the string object checking was creating the problem.

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+	var i;
+	for ( i = 0; i < str.length; i++ ) {
+		// Checking if a low surrogate is present at the beginning
+		if (i === 0 && RE_UTF16_LOW_SURROGATE.test(str[i])) {
```
```suggestion
		if ( i === 0 && RE_UTF16_LOW_SURROGATE.test(str[i]) ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js:
```diff
@@ -0,0 +1,79 @@
+		}
+
+		// Checking if a high surrogate is present at the last position
+		if (i === str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
```
```suggestion
		if ( i === str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i]) ) {
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/object.js:
```diff
@@ -0,0 +1,53 @@
+*/
+function isWellFormedString( str ) {
+	return (
+		isString.isObject(str) &&
```
```suggestion
		isString.isObject( str ) &&
```

Planeshifter commented on lib/node_modules/@stdlib/assert/is-well-formed-string/lib/object.js:
```diff
@@ -0,0 +1,53 @@
+function isWellFormedString( str ) {
+	return (
+		isString.isObject(str) &&
+		isWellFormed(str.valueOf())
```
```suggestion
		isWellFormed( str.valueOf() )
```

Planeshifter committed:
Apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented:
@gunjjoshi Will merge after CI clears; thanks again for your contribution!
